README
CurlFtpFS-NG 0.9.3-next

This is a fork of the unmaintained CurlFtpFS.  I don't have any great plans,
just fixing a bug that has already been fixed in various places and in various
ways, but now as a piece of software that has a name.

CurlFtpFS
=========

CurlFtpFS is a filesystem for acessing ftp hosts based on FUSE and
libcurl. It automatically reconnects if the server times out. 

My motivation to start this project was to learn how to program using
Curl (http://curl.haxx.se) and Fuse (http://fuse.sourceforge.net). I
also wanted to maintain my website (http://www.golaberto.com.br) using
the tools that I was used to, like cd, mv, cp and vim.

As the FTP protocol is not very feature rich, this filesystem does not
fulfill every constraint of a real filesystem, but it should be usable
for simple tasks like copying and editing files.

Requirements
============

glib-2.0
libcurl >= 7.17.0


Compilation and Installation
============================

./configure
make
make install

Usage
=====

curlftpfs <ftpsite> <mountpoint>

ex.:
curlftpfs ftp://ftp.sunet.se/ sunet/


Debugging
=========

curlftpfs -f -v -o debug,ftpfs_debug=3 <ftpsite> <mountpoint> 

(runs the curlftpfs in foreground and shows libcurl verbose debug output)  


Known Problems
==============

1) Several GUI application (gedit, leafpad,...) use open(O_RDWR) + seek mode for saving files 
which cannot be supported by the FTP protocol. Therefore saving files might throw an error.
Hopefully future kernels will provide special errno's to make it easier to 
deal with less capable file-systems.

2) There seems to be a bug in libcurl 7.18 which sometimes causes problems reading files. 
Should be fixed in libcurl >= 7.18.2

http://sourceforge.net/tracker/index.php?func=detail&aid=1951588&group_id=976&atid=100976






ChangeLog
--- CurlFtpFS-NG ---

0.9.3-next:
 * fix option `no_verify_hostname' with more recent curl versions

0.9.3:
 * migrate from cvs to git
 * fix filenames with URL-reserved characters
 * fix possible error on closing files after modifying
 * fix a memory leak
 * fix unit tests

--- CurlFtpFS ---

Apr 29, 2008
Reduce DEFAULT_CACHE_TIMEOUT from 300 to 10 seconds. List cache options and default timeout in --help.
mail2nob

Apr 29, 2008
release curlftpfs 0.9.2
mail2nob

Apr 29, 2008
new helper function cancel_previous_multi() which checks for errors.
mail2nob

Apr 25, 2008
bump configure.ac version 0.9.2; requires libcurl >=7.17.0
mail2nob

Apr 23, 2008 	
add Makefile.cvs 
mail2nob

Apr 23, 2008
patch by Miklos Szeredi: fixes 'du' for files >= 4GB. 
mail2nob

Jan 31, 2008
Fix missing -liconv.  	
robsonbraga

Nov 29, 2007
open(O_RDWR) --> fail with ENOTSUP  	
mail2nob

Nov 25, 2007
remove buffered writes completely and handle O_RDWR  	
mail2nob

Nov 20, 2007	
Fix "hang" when curl_easy_perform(fh->write_conn) in the write thread fails; Detect write errors in the write thread; Synchronize writing with the new semaphore "data_written" to be able to report write errors earlier; Fix that ftpfs_truncate created empty files with no permissions set (caused problems with gedit and others)
mail2nob

Nov 20, 2007	
Patch from Miklos Szeredi to support streaming writes (includes changes from Robson Braga Araujo)
mail2nob

Jul 26, 2007	
Added option to use a custom command to list files.
robsonbraga

Jun 28, 2007 	
Add ftp_method option to make it faster to browse directories.
robsonbraga

Jun 28, 2007	
Fix broken header dependencies.
robsonbraga

Jun 28, 2007	
Put the FTP server name in fsname by default. Upgraded FUSE_USE_VERSION to 26.
robsonbraga

Jun 14, 2007 	
Edited wiki page through web user interface.
robsonbraga

May 25, 2007 	
Disable EPSV by default because it can cause the program to hang.
robsonbraga

Apr 05, 2007 	
Edited wiki page through web user interface.
robsonbraga

Apr 05, 2007
Tagging the 0.9.1 release.
robsonbraga

Apr 05, 2007
Version 0.9.1.
robsonbraga

Apr 05, 2007
Revert changes from revisions 68 and 69 that enabled uploads even for servers that don't support overwrite. This creates a problem that we say the file is created but things like GETATTR don't see it in the server. We may want to make getattr look at local attributes in the future to make this work. 	
robsonbraga

Feb 28, 2007
Added option to be able to use SOCKS4/5 proxy. Only works with libcurl 7.16.2 or higher.
robsonbraga

Feb 27, 2007
Added debug levels.
robsonbraga

Feb 14, 2007
Generate a library so that tests don't have to compile files again.
robsonbraga

Feb 14, 2007
Remove warnings.
robsonbraga

Feb 13, 2007 	
Correct broken test.
robsonbraga

Feb 13, 2007
Corrected bug doing assignment instead of comparison.
robsonbraga

Feb 13, 2007
Added support for caching files not found, avoiding to do repeated access to the FTP server.
robsonbraga

Feb 10, 2007
Revert last change. FUSE_USE_VERSION is already defined in Makefile.am but old automake doesn't get it.
robsonbraga


tests/ftpfs-ls_unittest.c
/*
    FTP file system
    Copyright (C) 2006 Robson Braga Araujo <robsonbraga@gmail.com>

    This program can be distributed under the terms of the GNU GPL.
    See the file COPYING.
*/

#define _XOPEN_SOURCE 600  /* glibc2 needs this */
#include <time.h>                                                        
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>

#include "ftpfs.h"
#include "ftpfs-ls.h"

struct ftpfs ftpfs;

#define check(sbuf, dev, ino, mode, nlink, uid, gid, \
              rdev, size, blksize, blocks, date) \
  do { \
  struct tm tm; \
  time_t tt; \
  memset(&tm, 0, sizeof(tm)); \
  strptime(date, "%H:%M:%S %d/%m/%Y", &tm); \
  tt = mktime(&tm); \
  assert(sbuf.st_dev == (dev)); \
  assert(sbuf.st_ino == (ino)); \
  assert(sbuf.st_mode == (mode)); \
  assert(sbuf.st_nlink == (nlink)); \
  assert(sbuf.st_uid == (uid)); \
  assert(sbuf.st_gid == (gid)); \
  assert(sbuf.st_rdev == (rdev)); \
  assert(sbuf.st_size == (size)); \
  assert(sbuf.st_blksize == (blksize)); \
  assert(sbuf.st_blocks == (blocks)); \
  assert(sbuf.st_atime == tt); \
  assert(sbuf.st_ctime == tt); \
  assert(sbuf.st_mtime == tt); \
  } while (0);

int main(int argc, char **argv) {
  const char *list;
  char line[256];
  struct fuse_cache_operations dummy_oper;
  struct stat sbuf;
  int err;
  struct fuse_args args = FUSE_ARGS_INIT(argc, argv);
  char linkbuf[1024];
  char date[20];
  time_t tt;
  struct tm tm;
  struct tm test_tm;

  ftpfs.debug = 1;

  tt = time(NULL);
  gmtime_r(&tt, &tm);
  ftpfs.blksize = 4096;

  memset(&dummy_oper, 0, sizeof(dummy_oper));
  err = cache_parse_options(&args);
  cache_init(&dummy_oper);

  list = "05-22-03  12:13PM       <DIR>          chinese_pr\r\n";
  err = parse_dir(list, "/", "chinese_pr", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  check(sbuf, 0, 0, S_IFDIR, 1, 0, 0, 0, 0, 0, 0, "12:13:00 22/05/2003");

  err = parse_dir(list, "/", "hinese_pr", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 1);

  list = "05-14-03  02:49PM                40448 PR_AU13_CH.doc\r\n";
  err = parse_dir(list, "/", "PR_AU13_CH.doc", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  check(sbuf, 0, 0, S_IFREG, 1, 0, 0, 0, 40448, 4096, 80, "14:49:00 14/05/2003");

  list = "11-25-04  09:17AM             20075882 242_310_Condor_en_ok.pdf\r\n";
  err = parse_dir(list, "/", "242_310_Condor_en_ok.pdf", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  check(sbuf, 0, 0, S_IFREG, 1, 0, 0, 0, 20075882, 4096, 39216, "09:17:00 25/11/2004");

  list = "lrwxrwxrwx   1 1             17 Nov 24  2002 lg -> cidirb/documentos\r\n";
  err = parse_dir(list, "/", "lg", &sbuf, linkbuf, 1024, NULL, NULL);
  assert(err == 0);
  assert(!strcmp(linkbuf, "cidirb/documentos"));
  check(sbuf, 0, 0, S_IFLNK|S_IRUSR|S_IWUSR|S_IXUSR|S_IRGRP|S_IWGRP|S_IXGRP|S_IROTH|S_IWOTH|S_IXOTH, 1, 0, 0, 0, 17, 4096, 8, "00:00:00 24/11/2002");

  list = "lrwxrwxrwx    1 1137     1100           14 Mar 12  2004 molbio -> Science/molbio\r\n";
  err = parse_dir(list, "/", "molbio", &sbuf, linkbuf, 1024, NULL, NULL);
  assert(err == 0);
  assert(!strcmp(linkbuf, "Science/molbio"));
  check(sbuf, 0, 0, S_IFLNK|S_IRUSR|S_IWUSR|S_IXUSR|S_IRGRP|S_IWGRP|S_IXGRP|S_IROTH|S_IWOTH|S_IXOTH, 1, 0, 0, 0, 14, 4096, 8, "00:00:00 12/03/2004");

  // Test a date six months in the past
  test_tm = tm;
  test_tm.tm_mon -= 6;
  if (test_tm.tm_mon < 0) {
    test_tm.tm_mon += 12;
    test_tm.tm_year--;
  }
  strftime(line, 256,
           "drwxr-xr-x  4 robson users   4096 %b %d 00:00 tests\r\n", &test_tm);
  err = parse_dir(line, "/", "tests", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  strftime(date, 20, "00:00:00 %d/%m/%Y", &test_tm);
  check(sbuf, 0, 0, S_IFDIR|S_IRUSR|S_IWUSR|S_IXUSR|S_IRGRP|S_IXGRP|S_IROTH|S_IXOTH, 4, 0, 0, 0, 4096, 4096, 8, date);

  list = "dr-xr-xr-x   2 root     512 Apr  8  1994 etc\r\n";
  err = parse_dir(list, "/", "etc", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  check(sbuf, 0, 0, S_IFDIR|S_IRUSR|S_IXUSR|S_IRGRP|S_IXGRP|S_IROTH|S_IXOTH, 2, 0, 0, 0, 512, 4096, 8, "00:00:00 08/04/1994");

  // Test a date a little bit in the past
  test_tm = tm;
  strftime(line, 256,
           "----------  1 robson users   1803128 %b %d 00:00 ls-lR.Z\r\n",
           &test_tm);
  err = parse_dir(line, "/", "ls-lR.Z", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  strftime(date, 20, "00:00:00 %d/%m/%Y", &test_tm);
  check(sbuf, 0, 0, S_IFREG, 1, 0, 0, 0, 1803128, 4096, 3528, date);

  list = "-rw-r--r--  1 robson users   1803128 Jun 20 02:13  test\r\n";
  err = parse_dir(list, "/", " test", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  check(sbuf, 0, 0, S_IFREG|S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH, 1, 0, 0, 0, 1803128, 4096, 3528, "02:13:00 20/06/2006");

  list = "drwxrwsr-x+ 14 cristol molvis 1024 Feb 17 2000 v2\r\n";
  err = parse_dir(list, "/", "v2", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  check(sbuf, 0, 0, S_IFDIR|S_IRUSR|S_IWUSR|S_IXUSR|S_IRGRP|S_IWGRP|S_IXGRP|S_IROTH|S_IXOTH, 14, 0, 0, 0, 1024, 4096, 8, "00:00:00 17/02/2000");

  list = "drwxrwsr-x+144 cristol molvis 10240 Dec 31 2005 v11\r\n";
  err = parse_dir(list, "/", "v11", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  check(sbuf, 0, 0, S_IFDIR|S_IRUSR|S_IWUSR|S_IXUSR|S_IRGRP|S_IWGRP|S_IXGRP|S_IROTH|S_IXOTH, 144, 0, 0, 0, 10240, 4096, 24, "00:00:00 31/12/2005");

  list = "-rw-------    1 6700     2000     6561177600 Oct 15 2005 home.backup.tar\r\n";
  err = parse_dir(list, "/", "home.backup.tar", &sbuf, NULL, 0, NULL, NULL);
  assert(err == 0);
  check(sbuf, 0, 0, S_IFREG|S_IRUSR|S_IWUSR, 1, 0, 0, 0, 6561177600LL, 4096, 4426192, "00:00:00 15/10/2005");

  return 0;
}


cache.h
#ifndef __CURLFTPFS_CACHE_H__
#define __CURLFTPFS_CACHE_H__

/*
    Caching file system proxy
    Copyright (C) 2004  Miklos Szeredi <miklos@szeredi.hu>

    This program can be distributed under the terms of the GNU GPL.
    See the file COPYING.
*/

#include <fuse.h>
#include <fuse_opt.h>

#ifndef FUSE_VERSION
#define FUSE_VERSION (FUSE_MAJOR_VERSION * 10 + FUSE_MINOR_VERSION)
#endif

typedef struct fuse_cache_dirhandle *fuse_cache_dirh_t;
typedef int (*fuse_cache_dirfil_t) (fuse_cache_dirh_t h, const char *name,
                                    const struct stat *stbuf);

struct fuse_cache_operations {
    struct fuse_operations oper;
    int (*cache_getdir) (const char *, fuse_cache_dirh_t, fuse_cache_dirfil_t);
};

struct fuse_operations *cache_init(struct fuse_cache_operations *oper);
int cache_parse_options(struct fuse_args *args);
void cache_add_attr(const char *path, const struct stat *stbuf);
void cache_add_dir(const char *path, char **dir);
void cache_add_link(const char *path, const char *link, size_t size);

#endif   /* __CURLFTPFS_CACHE_H__ */


ftpfs.h
#ifndef __CURLFTPFS_FTPFS_H__
#define __CURLFTPFS_FTPFS_H__

/*
    FTP file system
    Copyright (C) 2006 Robson Braga Araujo <robsonbraga@gmail.com>

    This program can be distributed under the terms of the GNU GPL.
    See the file COPYING.
*/

#include <curl/curl.h>
#include <curl/easy.h>
#include <pthread.h>
#include <pthread.h>

struct ftpfs {
  char* host;
  char* mountpoint;
  pthread_mutex_t lock;
  CURL* connection;
  CURLM* multi;
  struct ftpfs_file* current_fh;
  unsigned blksize;
  int verbose;
  int debug;
  int transform_symlinks;
  int disable_epsv;
  int skip_pasv_ip;
  char* ftp_method;
  char* custom_list;
  int tcp_nodelay;
  char* ftp_port;
  int disable_eprt;
  int connect_timeout;
  int use_ssl;
  int no_verify_hostname;
  int no_verify_peer;
  char* cert;
  char* cert_type;
  char* key;
  char* key_type;
  char* key_password;
  char* engine;
  char* cacert;
  char* capath;
  char* ciphers;
  char* interface;
  char* krb4;
  char* proxy;
  int proxytunnel;
  int proxyanyauth;
  int proxybasic;
  int proxydigest;
  int proxyntlm;
  int proxytype;
  char* user;
  char* proxy_user;
  int ssl_version;
  int ip_version;
  char symlink_prefix[PATH_MAX+1];
  size_t symlink_prefix_len;
  curl_version_info_data* curl_version;
  int safe_nobody;
  int tryutf8;
  char *codepage;
  char *iocharset;
  int multiconn;
};

extern struct ftpfs ftpfs;

#define DEBUG(level, args...) \
        do { if (level <= ftpfs.debug) {\
               int i = 0; \
               while (++i < level) fprintf(stderr, " "); \
               fprintf(stderr, "%ld ", time(NULL));\
               fprintf(stderr, __FILE__ ":%d ", __LINE__);\
               fprintf(stderr, args);\
             }\
           } while(0)

#endif   /* __CURLFTPFS_FTPFS_H__ */


path_utils.c
/*
    FTP file system
    Copyright (C) 2007 Robson Braga Araujo <robsonbraga@gmail.com>

    This program can be distributed under the terms of the GNU GPL.
    See the file COPYING.
*/

#include "path_utils.h"
#include "charset_utils.h"
#include "ftpfs.h"

#include <string.h>
#include <stdlib.h>
#include <glib.h>

char* get_file_name(const char* path) {
  const char* filename = strrchr(path, '/');
  if (filename == NULL) filename = path;
  else ++filename;

  char* ret = strdup(filename);
  if (ftpfs.codepage) {
    convert_charsets(ftpfs.iocharset, ftpfs.codepage, &ret);
  }
  
  return ret;
}

char* get_full_path(const char* path) {
  char* ret;
  char* converted_path = NULL;
  
  ++path;

  if (ftpfs.codepage && strlen(path)) {
    converted_path = strdup(path);
    convert_charsets(ftpfs.iocharset, ftpfs.codepage, &converted_path);
    path = converted_path;
  }

  const char *const escaped_path = g_uri_escape_string(path, "/", FALSE);
  ret = g_strdup_printf("%s%s", ftpfs.host, escaped_path);

  free(converted_path);
  free((char *) escaped_path);

  return ret;
}

char* get_fulldir_path(const char* path) {
  char* ret;
  char* converted_path = NULL;

  ++path;

  if (ftpfs.codepage && strlen(path)) {
    converted_path = strdup(path);
    convert_charsets(ftpfs.iocharset, ftpfs.codepage, &converted_path);
    path = converted_path;
  }

  const char *const escaped_path = g_uri_escape_string(path, "/", FALSE);
  ret = g_strdup_printf(
    "%s%s%s", ftpfs.host, escaped_path, strlen(escaped_path) ? "/" : "");

  free(converted_path);
  free((char *) escaped_path);

  return ret;
}

char* get_dir_path(const char* path) {
  char* ret;
  char* converted_path = NULL;
  const char *lastdir;

  ++path;

  if (ftpfs.codepage) {
    converted_path = g_strdup(path);
    convert_charsets(ftpfs.iocharset, ftpfs.codepage, &converted_path);
    path = converted_path;
  }

  const char *const escaped_path = g_uri_escape_string(path, "/", FALSE);
  lastdir = strrchr(escaped_path, '/');
  if (lastdir == NULL) lastdir = escaped_path;

  ret = g_strdup_printf("%s%.*s%s",
                        ftpfs.host,
                        lastdir - escaped_path,
                        escaped_path,
                        lastdir - escaped_path ? "/" : "");

  free(converted_path);
  free((char *) escaped_path);

  return ret;
}


charset_utils.c
/*
    FTP file system
    Copyright (C) 2007 Robson Braga Araujo <robsonbraga@gmail.com>

    This program can be distributed under the terms of the GNU GPL.
    See the file COPYING.
*/

#include "config.h"

#include <string.h>
#include <limits.h>
#include <iconv.h>
#include <errno.h>
#include <stdlib.h>

#include "ftpfs.h"

int convert_charsets(const char* from, const char* to, char** str) {
  iconv_t cd;
  char* s = *str;

  if (!s || !*s)
    return 0;

  if (to && from && (cd = iconv_open(to, from)) != (iconv_t)-1) {
    ICONV_CONST char* ib;
    char* buf;
    char* ob;
    size_t ibl, obl;

    ibl = strlen(s) + 1;
    ib = s;
    obl = MB_LEN_MAX * ibl;
    buf = (char*)malloc(obl * sizeof(char));
    ob = buf;

    do {
      if (iconv(cd, &ib, &ibl, &ob, &obl) == (size_t)-1) {
        DEBUG("iconv return error %d\n", errno);
        if (obl) {
          *ob++ = *ib++;
          ibl--;
          obl--;
        }
      }
    } while (ibl && obl);
    *ob = 0;
    DEBUG("iconv return %s\n", buf);
    iconv_close (cd);
    free(*str);
    *str = buf;
  } else {
    DEBUG("iconv_open return error %d\n", errno);
  }
  
  return 0;
}


ftpfs-ls.c
/*
    FTP file system
    Copyright (C) 2006 Robson Braga Araujo <robsonbraga@gmail.com>

    This program can be distributed under the terms of the GNU GPL.
    See the file COPYING.
*/

#ifndef __FreeBSD__
#define _XOPEN_SOURCE 600
#else
#define _XOPEN_SOURCE
#endif

#include <time.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <glib.h>

#include "ftpfs.h"
#include "charset_utils.h"
#include "ftpfs-ls.h"

static int parse_dir_unix(const char *line,
                          struct stat *sbuf,
                          char *file,
                          char *link) {
  char mode[12];
  long nlink = 1;
  char user[33];
  char group[33];
  unsigned long long size;
  char month[4];
  char day[3];
  char year[6];
  char date[20];
  struct tm tm;
  time_t tt;
  int res;

  memset(file, 0, sizeof(char)*1024);
  memset(&tm, 0, sizeof(tm));
  memset(&tt, 0, sizeof(tt));

#define SPACES "%*[ \t]"
  res = sscanf(line,
               "%11s"
               "%lu"  SPACES
               "%32s" SPACES
               "%32s" SPACES
               "%llu" SPACES
               "%3s"  SPACES
               "%2s"  SPACES
               "%5s"  "%*c"
               "%1023c",
               mode, &nlink, user, group, &size, month, day, year, file);
  if (res < 9) {
    res = sscanf(line,
                 "%11s"
                 "%32s" SPACES
                 "%32s" SPACES
                 "%llu" SPACES
                 "%3s"  SPACES
                 "%2s"  SPACES
                 "%5s"  "%*c"
                 "%1023c",
                 mode, user, group, &size, month, day, year, file);
    if (res < 8) {
      return 0;
    }
  }
#undef SPACES

  char *link_marker = strstr(file, " -> ");
  if (link_marker) {
    strcpy(link, link_marker + 4);
    *link_marker = '\0';
  }

  int i = 0;
  if (mode[i] == 'd') {
    sbuf->st_mode |= S_IFDIR;
  } else if (mode[i] == 'l') {
    sbuf->st_mode |= S_IFLNK;
  } else {
    sbuf->st_mode |= S_IFREG;
  }
  for (i = 1; i < 10; ++i) {
    if (mode[i] != '-') {
      sbuf->st_mode |= 1 << (9 - i);
    }
  }

  sbuf->st_nlink = nlink;

  sbuf->st_size = size;
  if (ftpfs.blksize) {
    sbuf->st_blksize = ftpfs.blksize;
    sbuf->st_blocks =
      ((size + ftpfs.blksize - 1) & ~(ftpfs.blksize - 1)) >> 9;
  }

  sprintf(date,"%s,%s,%s", year, month, day);
  tt = time(NULL);
  gmtime_r(&tt, &tm);
  tm.tm_sec = tm.tm_min = tm.tm_hour = 0;
  if(strchr(year, ':')) {
    int cur_mon = tm.tm_mon;  // save current month
    strptime(date, "%H:%M,%b,%d", &tm);
    // Unix systems omit the year for the last six months
    if (cur_mon + 5 < tm.tm_mon) {  // month from last year
      DEBUG(2, "correct year: cur_mon: %d, file_mon: %d\n", cur_mon, tm.tm_mon);
      tm.tm_year--;  // correct the year
    }
  } else {
    strptime(date, "%Y,%b,%d", &tm);
  }

  sbuf->st_atime = sbuf->st_ctime = sbuf->st_mtime = mktime(&tm);

  return 1;
}

static int parse_dir_win(const char *line,
                         struct stat *sbuf,
                         char *file,
                         char *link) {
  char date[9];
  char hour[8];
  char size[33];
  struct tm tm;
  time_t tt;
  int res;
  (void)link;

  memset(file, 0, sizeof(char)*1024);
  memset(&tm, 0, sizeof(tm));
  memset(&tt, 0, sizeof(tt));

  res = sscanf(line, "%8s%*[ \t]%7s%*[ \t]%32s%*[ \t]%1023c",
               date, hour, size, file);
  if (res < 4) {
    return 0;
  }

  DEBUG(2, "date: %s hour: %s size: %s file: %s\n", date, hour, size, file);

  tt = time(NULL);
  gmtime_r(&tt, &tm);
  tm.tm_sec = tm.tm_min = tm.tm_hour = 0;
  strptime(date, "%m-%d-%y", &tm);
  strptime(hour, "%I:%M%p", &tm);

  sbuf->st_atime = sbuf->st_ctime = sbuf->st_mtime = mktime(&tm);

  sbuf->st_nlink = 1;

  if (!strcmp(size, "<DIR>")) {
    sbuf->st_mode |= S_IFDIR;
  } else {
    unsigned long long nsize = strtoull(size, NULL, 0);
    sbuf->st_mode |= S_IFREG;
    sbuf->st_size = nsize;
    if (ftpfs.blksize) {
      sbuf->st_blksize = ftpfs.blksize;
      sbuf->st_blocks =
        ((nsize + ftpfs.blksize - 1) & ~(ftpfs.blksize - 1)) >> 9;
    }
  }

  return 1;
}

static int parse_dir_netware(const char *line,
                             struct stat *sbuf,
                             char *file,
                             char *link) {
  (void) line;
  (void) sbuf;
  (void) file;
  (void) link;
  return 0;
}


int parse_dir(const char* list, const char* dir,
              const char* name, struct stat* sbuf,
              char* linkbuf, int linklen,
              fuse_cache_dirh_t h, fuse_cache_dirfil_t filler) {
  char *file;
  char *link;
  const char *start = list;
  const char *end = list;
  char found = 0;
  struct stat stat_buf;

  if (sbuf) memset(sbuf, 0, sizeof(struct stat));

  if (name && sbuf && name[0] == '\0') {
    sbuf->st_mode |= S_IFDIR;
    sbuf->st_mode |= 0755;
    sbuf->st_size = 1024;
    sbuf->st_nlink = 1;
    return 0;
  }

  file = (char *)malloc(1024*sizeof(char));
  link = (char *)malloc(1024*sizeof(char));

  while ((end = strchr(start, '\n')) != NULL) {
    char* line;

    memset(&stat_buf, 0, sizeof(stat_buf));

    if (end > start && *(end-1) == '\r') end--;

    line = (char*)malloc(end - start + 1);
    strncpy(line, start, end - start);
    line[end - start] = '\0';
    start = *end == '\r' ? end + 2 : end + 1;

    if (ftpfs.codepage) {
      convert_charsets(ftpfs.codepage, ftpfs.iocharset, &line);
    }

    file[0] = link[0] = '\0';
    int res = parse_dir_unix(line, &stat_buf, file, link) ||
              parse_dir_win(line, &stat_buf, file, link) ||
              parse_dir_netware(line, &stat_buf, file, link);

    if (res) {
      char *full_path = g_strdup_printf("%s%s", dir, file);

      if (link[0]) {
        char *reallink;
        if (link[0] == '/' && ftpfs.symlink_prefix_len) {
          reallink = g_strdup_printf("%s%s", ftpfs.symlink_prefix, link);
        } else {
          reallink = g_strdup(link);
        }
        int linksize = strlen(reallink);
        cache_add_link(full_path, reallink, linksize+1);
        DEBUG(1, "cache_add_link: %s %s\n", full_path, reallink);
        if (linkbuf && linklen) {
          if (linksize > linklen) linksize = linklen - 1;
          strncpy(linkbuf, reallink, linksize);
          linkbuf[linksize] = '\0';
        }
        free(reallink);
      }

      if (h && filler) {
        DEBUG(1, "filler: %s\n", file);
        filler(h, file, &stat_buf);
      } else {
        DEBUG(1, "cache_add_attr: %s\n", full_path);
        cache_add_attr(full_path, &stat_buf);
      }

      DEBUG(2, "comparing %s %s\n", name, file);
      if (name && !strcmp(name, file)) {
        if (sbuf) *sbuf = stat_buf;
        found = 1;
      }

      free(full_path);
    }

    free(line);
  }

  free(file);
  free(link);

  return !found;
}
